pwiz_tools/Skyline/Model/Koina/Models/KoinaModel.cs

```diff
@@ -705,26 +705,79 @@ public static double CalculateSpectrumDotpMzMatch(LibraryRankedSpectrumInfo spec
                 new Statistics(intensities2All));
         }
 
-        /*public static double CalculateSpectrumDotpIonMatch(LibraryRankedSpectrumInfo spectrum1,
-            LibraryRankedSpectrumInfo spectrum2)
+        /// <summary>
+        /// Calculates the normalized contrast angle of the square rooted
+        /// intensities of the two given spectra using all peaks and possibly ignoring
+        /// mismatched peaks
+        /// </summary>
+        /// <param name="spectrum1">First spectrum</param>
+        /// <param name="spectrum2">Second spectrum</param>
+        /// <param name="mzTolerance">Tolerance for considering two mzs the same</param>
+        /// <param name="ignoreMismatches1">if true, unmatched peaks from the first spectrum are ignored</param>
+        /// <param name="ignoreMismatches2">if true, unmatched peaks from the second spectrum are ignored</param>
+        public static double CalculateSpectrumDotpMzFull(LibraryRankedSpectrumInfo spectrum1,
+            LibraryRankedSpectrumInfo spectrum2, MzTolerance mzTolerance, bool ignoreMismatches1, bool ignoreMismatches2)
         {
-            var matched1 = spectrum1.PeaksMatched.ToArray();
-            var matched2 = spectrum2.PeaksMatched.ToArray();
-            var intensities1All = new List<double>(matched1.Length + matched2.Length);
-            var intensities2All = new List<double>(matched1.Length + matched2.Length);
-
-            foreach (var match1 in matched1)
+            var vectorLength = spectrum1.Intensities.Count() + spectrum2.Intensities.Count();
+            var s1 = Enumerable.Zip(spectrum1.MZs, spectrum1.Intensities,

```

Comment:
This is almost the same as `s1=spectrum1.Peaks.ToList()`, except that the code here produces a List of MI whereas "Peaks" is a list of RankedMI.